File: Python/004-largest-palindrome-product.py
def isPalin(num):
    '''Checks if a number is a palindrome'''
    num = list(str(num))
    if num == num[::-1]:
        return True
    return False

def maxPalin(n):
    '''Largest product of two numbers that's a palindrome of length n''' 
    smallest = 10**(n-1) if n > 1 else 0 #each n digit number starts here, 1 dig nums start at 0, 2 dig nums start at 10^(2-1), 3 digs at 10^(3-1)
    biggest =  10**(n) #last n length number to check for, for 3 digs, 1000-1 = 999, loop goes from smallest to biggest-1
    maxNum = 0

    for i in range(smallest, biggest):
        for j in range(smallest, biggest):
            if isPalin(i*j) and (i*j) > maxNum:
                maxNum = i*j
    return maxNum

print(maxPalin(2)) #should return 9009 
print(maxPalin(3)) #returns 906609

File: Ruby/004-largest-palindrome-product.rb
# A palindromic number reads the same both ways. 
# The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

# Find the largest palindrome made from the product of two 3-digit numbers.

def is_palindrome(number) 
  number_string = number.to_s
  len = number_string.length
  character_matches_mirror = -> (i) { number_string[i] === number_string[len - i - 1]}
  for i in 0...(len / 2)
    return false unless character_matches_mirror.(i)
  end
  true
end

def largest_palindrome_product(top = 1000)
  best = 0
  # check each combination of two numbers 
  for i in 100...(top)
    for j in i...(top)
      k = i*j
      best = k if k > best && is_palindrome(k)
    end
  end
  best
end

puts largest_palindrome_product